org.scala-ide.sdt.core/src/org/scalaide/extensions/AutoEdit.scala

```diff
@@ -0,0 +1,117 @@
+package org.scalaide.extensions
+
+import org.scalaide.core.text.Change
+import org.scalaide.core.text.TextChange
+
+/**
+ * Base interface for auto edits. An auto edit is an IDE extension that is
+ * executed while an user types.
+ */
+trait AutoEdit extends ScalaIdeExtension with DocumentSupport {
+
+  /**
+   * Contains the text change the user did to the document. In case the user
+   * inserted or removed a character, the text change has a range size of one.
+   * In case the user removed a selection or pasted text into the editor the
+   * text change can have a range of any size.
+   *
+   * This value must not be overwritten at the location where [[setting()]] or
+   * [[perform()]] are implemented. Instead it is implemented by
+   * the IDE at the location where the auto edit is instantiated.
+   */
+  val textChange: TextChange
+
+  /**
+   * The setting information is used in the "Scala > Editor > Auto Edits"
+   * preference page of Eclipse. On this preference page users can enable or
+   * disable auto edits and add further configuration to them. If the auto edit
+   * is disabled, [[perform()]] will never be executed.
+   */
+  override def setting: AutoEditSetting
+
+  /**
+   * Performs the auto edit and returns an instance of [[Change]] when the auto
+   * edit should be applied to the document. In case the auto edit couldn't be
+   * applied, [[None]] needs to be returned.
+   *
+   * In the case that [[None]] is returned the next auto edit is performed, if
+   * one exists. If a [[Some]] is returned, no further auto edits will be
+   * performed.
+   *
+   * This method will never be called when this auto edit is disabled.
+   */
+  def perform(): Option[Change]
+
+  /**
+   * Checks if `pf` is defined for `a` and returns the result of `pf` if this is
+   * the case. Returns [[None]] otherwise.
+   */
+  final def check[A, B](a: A)(pf: PartialFunction[A, Option[B]]): Option[B] =
+    if (pf.isDefinedAt(a)) pf(a) else None
+
```

Comment:
I let them as is, since these functions on `PartialFunction` have some runtime overhead and they are not really easier to read.